### PR TITLE
Fix nim-lang/nimforum#285 - punctuation after URL

### PR DESCRIPTION
**Pull request by a-mr**
e.g. in such a link
```
(http://no.org),
```
final `)` was included into URL, which contradicts to RST spec and Markdown behavior.

### CODE REVIEW

**Comment by Araq** [comment]
Ah finally! That drove me nuts...

File: tests/stdlib/trst.nim [hunk 1]
```diff
@@ -428,3 +428,46 @@ suite "RST inline markup":
...
+    check(dedent"""
+        [see (http://no.org)], end""".toAst ==
+      dedent"""
+        rnInner
+          rnLeaf  '['
+          rnLeaf  'see'
+          rnLeaf  ' '
+          rnLeaf  '('
+          rnStandaloneHyperlink
+            rnLeaf  'http'
```
Comment (timotheecour) [inline tests/stdlib/trst.nim]: not blocking this PR but why isn't the whole url a single leaf node?

Comment (a-mr) [reply tests/stdlib/trst.nim]: pre-existing way:-)
agree, changed to single

File: tests/stdlib/trst.nim [hunk 2]
```diff
@@ -439,12 +439,7 @@ suite "RST inline markup":
           rnLeaf  ' '
           rnLeaf  '('
           rnStandaloneHyperlink
-            rnLeaf  'http'
-            rnLeaf  ':'
-            rnLeaf  '//'
-            rnLeaf  'no'
-            rnLeaf  '.'
-            rnLeaf  'org'
+            rnLeaf  'http://no.org'
```
Comment (timotheecour) [inline tests/stdlib/trst.nim:442]: in this or followup PR, can you add tests with more complicated urls, involving ?, & etc ? or are there already tests for those?

Comment (a-mr) [reply tests/stdlib/trst.nim:442]: Added some arbitrary link https://www.google.com/url?sa=t&source=web&cd=&cad=rja&url=https%3A%2F%2Fnim-lang.github.io%2FNim%2Frst.html%23features&usg=AO&= 

Addition of final `&=` is invented to check the current behavior. Previously it would include `&` but not final `=`. Now it skips both `&=`.

File: tests/stdlib/trst.nim [hunk 3]
```diff
@@ -460,3 +460,21 @@ suite "RST inline markup":
           rnLeaf  ' '
           rnLeaf  'end'
         """)
+
+    # a more complex URL with some made-up ending '&='.
+    # Github Markdown would include final &= and
+    # so would rst2html.py in contradiction with RST spec.
```
Comment (timotheecour) [inline tests/stdlib/trst.nim:466]: do you have a link discussing this?

it's such an edge case that it's good enough for this PR, but would be good to link to spec since there's a conflict between commonly used tools and spec

Comment (a-mr) [reply tests/stdlib/trst.nim:466]: Only https://docutils.sourceforge.io/docs/ref/rst/restructuredtext.html#standalone-hyperlinks
> Punctuation at the end of a URI is not considered part of the URI, unless the URI is terminated by a closing angle bracket (">").

Punctuation characters are not strictly defined but RST spec mentions `&` and `=` among them in https://docutils.sourceforge.io/docs/ref/rst/restructuredtext.html#sections . It also mentions Unicode "punctuation" category (though `=` does not fall into it)

And `tkPunct` in `rst.nim` has its own list of characters.

Interestingly, vim syntax highlighting parses that string just as this PR.

I think such a case is very rare because URLs almost always have format like `?=word` or `#word` so I would not bother to dig more.

Comment (a-mr) [reply tests/stdlib/trst.nim:466]: BTW `)` and other punctuation characters are allowed in URLs generally. So the mentioned rule is an invention of those markup language for the sake of convenience.

rst2html.py claims that it parses URLs according to the corresponding RFC for each protocol type. (Of course we don't implement that stuff.) Though its behavior seems weird in some cases:

image

Github Markdown is more reasonable here:

See https://www.google.com/url?sa=t&source=web&cd=&cad=rja&url=https%3A%2F%2Fnim-lang.github.io%2FNim%2Frst.html%23features&usg=AO&  end.

I just don't feel like it's a good time investment to dig into all those minor edge cases.

Comment (timotheecour) [reply tests/stdlib/trst.nim:466]: thanks for the links, and yes, it's good enough :)
